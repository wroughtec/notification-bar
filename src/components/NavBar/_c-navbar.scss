.c-navbar {
  position: relative;

  &__container {
    display: flex;
    height: var(--navbarHeight);
    padding-top: .5rem;

    padding-right: 1.5rem;
    padding-bottom: .5rem;
    padding-left: 1.5rem;

    justify-content: space-between;
    align-items: center;

    color: var(--navbarColour);

    background-color: var(--navbarBackground);
  }

  &__title {
    margin-top: 0;
    margin-bottom: 0;

    font-size: var(--navbarFontSize);
    font-weight: 400;
  }

  &__btn {
    position: relative;

    display: flex;

    padding: .75rem;
    overflow: visible;

    font-size: 1.5rem;
    color: var(--navbarColour);
    text-align: center;

    cursor: pointer;

    background-color: transparent;
    border: 0;
    border-radius: 50%;

    transition: background-color 150ms cubic-bezier(.4, 0, .2, 1) 0ms;
  }

  &__icon {
    font-size: var(--navbarFontSize);
  }
}
