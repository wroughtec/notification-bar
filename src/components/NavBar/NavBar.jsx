// @flow
import React, { Component } from 'react';
import { Icon } from 'components/Icon/Icon';
import { user } from 'consts/iconIds';
import { asyncFetch } from 'utils/asyncFetch';
import { Menu } from 'components/Menu/Menu';
import { Badge } from 'components/Badge/Badge';
import './_c-navbar.scss';

type State = {
  notifications: NotificationType[],
  unreadNotifications: number,
  loading: boolean,
  isMenuOpen: boolean,
  clearNotifications: boolean
};

export class NavBar extends Component<void, State> {
  state = {
    notifications: [],
    unreadNotifications: 0,
    loading: true,
    isMenuOpen: false,
    clearNotifications: false
  };

  async componentDidMount() {
    const getNotificaTions = await asyncFetch();

    this.addNotifcation(getNotificaTions);
  }

  addNotifcation: Function = (notifications: NotificationType[]) => {
    this.setState({
      notifications,
      unreadNotifications: notifications.length,
      loading: false
    });
  };

  onClick: Function = (event: SyntheticEvent<HTMLButtonElement>) => {
    event.preventDefault();
    const { isMenuOpen, clearNotifications } = this.state;

    if (!clearNotifications) {
      this.setState({
        clearNotifications: true
      });
    } else {
      this.setState({ unreadNotifications: 0 });
    }

    if (!isMenuOpen) {
      this.setState(
        {
          isMenuOpen: true
        },
        () => {
          document.addEventListener('click', this.closeMenu);
        }
      );
    }
  };

  closeMenu: Function = () => {
    this.setState({ isMenuOpen: false }, () => {
      document.removeEventListener('click', this.closeMenu);
    });
  };

  render() {
    const { notifications, unreadNotifications, loading, isMenuOpen } = this.state;

    return (
      <nav className="c-navbar">
        <div className="c-navbar__container">
          <h2 className="c-navbar__title">Notification Bar</h2>
          {!loading && (
            <button type="button" className="c-navbar__btn" onClick={this.onClick}>
              <Icon iconId={user} iconClass="c-navbar__icon" />
              <Badge unreadNotifications={unreadNotifications} />
            </button>
          )}
        </div>
        {isMenuOpen && <Menu notifications={notifications} />}
      </nav>
    );
  }
}
