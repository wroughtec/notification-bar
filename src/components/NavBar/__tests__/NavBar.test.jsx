import fetchMock from 'fetch-mock';
import React from 'react';
import { shallow } from 'enzyme';
import { NavBar } from 'components/NavBar/NavBar';
import mockData from 'mocks/test-mocky.json';

fetchMock.get('*', JSON.stringify([...mockData]));

describe('NavBar', () => {
  let component;

  beforeEach(async () => {
    component = await shallow(<NavBar />);
  });

  it('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('loading has complete', () => {
    expect(component.state().loading).toEqual(false);
  });

  it('notification data should be in local state', () => {
    expect(component.state().notifications).toMatchObject([...mockData]);
  });

  it('notification length should equal 6', () => {
    expect(component.state().unreadNotifications).toEqual(6);
  });

  it('menu state is open when button clicked', async () => {
    await component.find('button').simulate('click', { preventDefault() {} });

    expect(component.state().isMenuOpen).toEqual(true);
    expect(component.state().clearNotifications).toEqual(true);
  });

  it('unreadNotifications are cleared when button clicked twice', async () => {
    await component.find('button').simulate('click', { preventDefault() {} });
    await component.find('button').simulate('click', { preventDefault() {} });

    expect(component.state().unreadNotifications).toEqual(0);
  });
});
