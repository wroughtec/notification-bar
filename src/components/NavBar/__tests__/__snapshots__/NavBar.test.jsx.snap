// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NavBar />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="c-navbar__container"
        >
          <h2
            className="c-navbar__title"
          >
            Notification Bar
          </h2>
          <button
            className="c-navbar__btn"
            onClick={[Function]}
            type="button"
          >
            <Icon
              iconClass="c-navbar__icon"
              iconId="user"
            />
            <Badge
              unreadNotifications={6}
            />
          </button>
        </div>,
        false,
      ],
      "className": "c-navbar",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2
              className="c-navbar__title"
            >
              Notification Bar
            </h2>,
            <button
              className="c-navbar__btn"
              onClick={[Function]}
              type="button"
            >
              <Icon
                iconClass="c-navbar__icon"
                iconId="user"
              />
              <Badge
                unreadNotifications={6}
              />
            </button>,
          ],
          "className": "c-navbar__container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Notification Bar",
              "className": "c-navbar__title",
            },
            "ref": null,
            "rendered": "Notification Bar",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Icon
                  iconClass="c-navbar__icon"
                  iconId="user"
                />,
                <Badge
                  unreadNotifications={6}
                />,
              ],
              "className": "c-navbar__btn",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "iconClass": "c-navbar__icon",
                  "iconId": "user",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "unreadNotifications": 6,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "button",
          },
        ],
        "type": "div",
      },
      false,
    ],
    "type": "nav",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="c-navbar__container"
          >
            <h2
              className="c-navbar__title"
            >
              Notification Bar
            </h2>
            <button
              className="c-navbar__btn"
              onClick={[Function]}
              type="button"
            >
              <Icon
                iconClass="c-navbar__icon"
                iconId="user"
              />
              <Badge
                unreadNotifications={6}
              />
            </button>
          </div>,
          false,
        ],
        "className": "c-navbar",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2
                className="c-navbar__title"
              >
                Notification Bar
              </h2>,
              <button
                className="c-navbar__btn"
                onClick={[Function]}
                type="button"
              >
                <Icon
                  iconClass="c-navbar__icon"
                  iconId="user"
                />
                <Badge
                  unreadNotifications={6}
                />
              </button>,
            ],
            "className": "c-navbar__container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Notification Bar",
                "className": "c-navbar__title",
              },
              "ref": null,
              "rendered": "Notification Bar",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Icon
                    iconClass="c-navbar__icon"
                    iconId="user"
                  />,
                  <Badge
                    unreadNotifications={6}
                  />,
                ],
                "className": "c-navbar__btn",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "iconClass": "c-navbar__icon",
                    "iconId": "user",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "unreadNotifications": 6,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "button",
            },
          ],
          "type": "div",
        },
        false,
      ],
      "type": "nav",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
